fs-index/src/watch.rs

```diff
@@ -0,0 +1,139 @@
+        root_path.as_ref()
+    );
+
+    let root_path = fs::canonicalize(root_path.as_ref())?;
+    let mut index: ResourceIndex<Id> = ResourceIndex::build(&root_path)?;
+    index.store()?;
+
+    let (mut watcher, mut rx) = async_watcher()?;
+    info!("Watching directory: {:?}", root_path);
+    let config = Config::default();
+    watcher.configure(config)?;
+    watcher.watch(root_path.as_ref(), RecursiveMode::Recursive)?;
+    info!("Started watcher with config: \n\t{:?}", config);
+
+    let ark_folder = root_path.join(ARK_FOLDER);
+    while let Some(res) = rx.next().await {
+        match res {
+            Ok(event) => {
+                // If the event is a change in .ark folder, ignore it
+                if event
+                    .paths
+                    .iter()
+                    .any(|p| p.starts_with(&ark_folder))
+                {
+                    continue;
+                }
+                // We only care for:
+                // - file modifications
+                // - file renames
+                // - file creations
+                // - file deletions
+                match event.kind {
+                    notify::EventKind::Modify(
+                        notify::event::ModifyKind::Data(_),
+                    )
+                    | notify::EventKind::Modify(
+                        notify::event::ModifyKind::Name(_),
+                    )
+                    | notify::EventKind::Create(
+                        notify::event::CreateKind::File,
+                    )
+                    | notify::EventKind::Remove(
+                        notify::event::RemoveKind::File,
+                    ) => {}
+                    _ => continue,
+                }
+
+                // If the event requires a rescan, update the entire index
+                // else, update the index for the specific file
+                if event.need_rescan() {
+                    info!("Detected rescan event: {:?}", event);
+                    index.update_all()?;
+                } else {
+                    info!("Detected event: {:?}", event);
+                    let file = event
+                        .paths
+                        .first()
+                        .expect("Failed to get file path from event");
+                    log::debug!("Updating index for file: {:?}", file);
+
+                    log::info!(
+                        "\n Current resource index: {}",
+                        index
+                            .resources()
+                            .iter()
+                            .map(|x| x.path().to_str().unwrap().to_string())
+                            .collect::<Vec<String>>()
+                            .join("\n\t")
+                    );
+
+                    let relative_path = file.strip_prefix(&root_path)?;
+                    log::info!("Relative path: {:?}", relative_path);
+                    index.update_one(relative_path)?;
```

Comment:
The result should be used to provide user with actual updates.

Reply:
I chose the first approach to get things running faster and keep it simpler for now. I also plan to add tests soon.

Next, I want to add integration tests for this functionality. We could either add integration tests for `fs-index` directly or implement CI shell scripts to test `ark-cli watch <PATH>` for an end-to-end approach—possibly both?

Do you think CI shell scripts for `ark-cli watch` would be sufficient, or should we also include programmatic tests? For example, running the watcher in a separate thread and doing many create/delete operations to verify the results. Now that I think about it, writing these tests programmatically could be complex. What’s your take?

Reply:
> Do you think CI shell scripts for ark-cli watch would be sufficient, or should we also include programmatic tests? For example, running the watcher in a separate thread and doing many create/delete operations to verify the results. Now that I think about it, writing these tests programmatically could be complex. What’s your take?

Agree, I think we can achieve proper result by simple shell script. I imagine it like this:
1. Run `ark-cli watch` in background, direct its output to dedicated log file.
2. Randomly modify folder content and write the performed actions into another log file.
3. Then compare the two log files.

Reply:
Added a shell script `integration/ark-cli-watch.sh` to check the sanity of `ark-cli watch`. I think it’s pretty cool because it also checks other parts of the code for sanity along the way, like `update_one` in this case. Using `ark-cli` is definitely a great way to write end-to-end shell scripts to verify what we have.
  
I’ve also set it up to run in the CI with each push/PR to the main branch. You can check out the expected workflow in [my fork here](https://github.com/tareknaser/ark-rust/actions/runs/11576918354). I’ve been using it for debugging